### ISSUE TITLE
CLIP features still aligned with text?

### ISSUE THREAD

**Issue by gtziafas** [mhamilton723/FeatUp#6]
Hey, thanks for this awesome project! I use the CLIP upsampler out-of-the-box to get pixel features, but if I then compute cosine similarity with text embeddings (given from same model via `upsampler.mode[0].model.encode_text`), it seems that the alignment between image-text features is gone (check screenshot). 

![Figure11](https://github.com/mhamilton723/FeatUp/assets/23073814/7813189b-ebc5-4ddf-a747-d96d6823ba6a)

Am I doing something wrong or this is expected?

The code is:
```python
image_tensor = ... 
text = ...

upsampler = torch.hub.load("mhamilton723/FeatUp", 'clip').cuda()

vis_feats = upsampler(image_tensor.cuda())
txt_feats = upsampler.model[0].model.encode_text( tokenize(text).cuda() )
vis_feats /= vis_feats.norm(dim=-1, keedim=True)
txt_feats /= txt_feats.norm(dim=-1, keepdim=True)
sims = vis_feats @ txt_feats.T
```

**Comment by mhamilton723** [comment]
Hey @gtziafas some of this might be due to the layer norm operation we use in the pretrained models so that their features arent all over the place in terms of magnitude prior to upsampling. We'll be releasing some non-layer-norm pretrained models soon so that they can truly be drop in.

**Comment by wb014** [comment]
@mhamilton723 Thanks for the work.
In my experiment, the alignment is gone also when ‘use_norm’ set to False.
Left one is the similarity between hr_feats and text feature of 'a cat'.
<img width="1863" alt="cat" src="https://github.com/mhamilton723/FeatUp/assets/53701627/f0c437b9-4ed5-4438-8057-ede603afcbcf">

**Comment by gtziafas** [comment]
Hey @mhamilton723, thanks for the fast update! I load the upsampler with 
`upsampler = torch.hub.load("mhamilton723/FeatUp", 'clip', use_norm=False).cuda()`

It seems that the image-text similarity has similar behavior than before 
![temp_CLIPJUB](https://github.com/mhamilton723/FeatUp/assets/23073814/90e70749-6b75-4e08-b1ef-f578588b4e25)

However, interestingly, I see that if I simply reverse the normalized similarities:
```python
sims = vis_feats @ txt_feats.T
sims_norm = (sims - sims.min()) / (sims.max() - sims.min())
sims_norm_rev = 1.0 - sims_norm
```
there are some signs of life:

![temp_CLIPJUB_1](https://github.com/mhamilton723/FeatUp/assets/23073814/e673cc62-314d-419e-9e72-fbffd4476ba0)

I' m not sure where this behavior comes from, some numerical conditioning in the upsampler? I will try to find the time to  test the quality of the reversed normalized upsampler embeddings vs. standard CLIP patch embeddings for same pretrained model, e.g. from things like MaskCLIP

**Comment by mhamilton723** [comment]
@gtziafas @wb014 thanks for working on this. I would first do a sanity check that the CLIP model itself has the property that you want in the LR features because without that we have no hope. 

This is my clip featurizer code:
https://github.com/mhamilton723/FeatUp/blob/main/featup/featurizers/CLIP.py

And its referencing this function i added to the base clip model:
https://github.com/mhamilton723/CLIP/blob/dce9d43cc659cb9a013745b378c78077fde6aac0/clip/model.py#L403

If this looks fishy to you please let me know and ill remake the CLIP models!

**Comment by wb014** [comment]
@mhamilton723 Maybe in fact CLIP does not have the alignment between spatial feature and text feature. And I test MaskCLIP, which use 'value' of the last layer's attention module as dense feature map. The alignment can be seen to some extent.
<img width="973" alt="截屏2024-03-28 15 32 39" src="https://github.com/mhamilton723/FeatUp/assets/53701627/f99c1c23-154b-4a15-9645-fb6589e3cc09">

Do you have any idea of training an upsampler based on MaskCLIP? I can share you the MaskCLIP code, only adding some funtions based on the origin CLIP.

**Comment by gtziafas** [comment]
@mhamilton723  @wb014  Indeed using the MaskCLIP way of obtaining patch features maintains the text aligment.

This is from **low-res MaskCLIP ViT-B/16** (the queries are: "skateboard",  "yellow jacket")

![Selection_448](https://github.com/mhamilton723/FeatUp/assets/23073814/cd74cce8-ed23-4532-9031-313092c6e7f5)


I integrated MaskCLIP as a new backbone and trained an implicit upsampler on this scene. The results show promise, the upscaled patch features have more crisp alignment with text

![Selection_447](https://github.com/mhamilton723/FeatUp/assets/23073814/e57d77dd-8007-4289-b4dc-db2e6bdbcfb9)

Results with **high-res MaskCLIP ViT-B/16 backbone + implicit upsampler**:
![Selection_449](https://github.com/mhamilton723/FeatUp/assets/23073814/7feb2194-1010-49f5-9da7-34279939571a)

**Comment by mhamilton723** [comment]
THis is awesome and looking forward to playing with this. I think the implicit results look a little fuzzy and once we get this in ill do some playing with them to try to improve crisp-ness

**Comment by wb014** [comment]
@mhamilton723 @gtziafas I see the MaskCLIP commit, thanks for your effort! May I ask whether the training code of JBU upsample be released in few days? Or is there any plan of training MaskCLIP JBU? I'm waiting for this in anxiety.

**Comment by mhamilton723** [comment]
@wb014 trying to get this training now, i still need to make some edits to the custom JBU training in this repo before it will work the way it does to get the high quality backbones we released. I will add MaskCLIP to the hub in a few days

**Comment by mhamilton723** [comment]
@wb014 @gtziafas just added the JBU upsampler for MaskCLIP. Huge thanks to @gtziafas for the effort here. Also if anyone would like to contribute a quick example to the collab of getting upsampled heatmaps would certainly approve :)
